### ISSUE TITLE
Is there a way to specify a Vue component inside of the columns declaration?

### ISSUE THREAD

**Issue by ajmcgrail** [DataTables/Vue#38]
Hi there. Big fan of the change that now allows you to use Vue components in named slots, but I had a question/feature request (not sure if it's already implemented). Is there a way to specify a Vue component that will be used WITHOUT using a named slot?

For example, let's say that I want to have a list of button columns that I may or may not want to render, or I want to change the order on them. The point is that using a named slot wouldn't easily allow that. Is there a way to instead pass the Vue component directly in the defined columns of the DataTable component props?

I imagine it would look something like this:

```
// MyVueComponent is the just the imported vue component
const things: ConfigColumns[] = [{ data: null, render: MyVueComponent, additionalProps: {testProp: 5} }];
```

But as far as I can tell this isn't supported? I took at look at the DataTable Vue component source code and I think it looks feasible to implement with the existing logic of how Vue components are rendered, along with passing additional props. I might be able to try what I want in a PR if that would be preferred.

Please let me know if I need to give any more information. Thanks for your time!

**Comment by AllanJard** [comment]
Hi - at the moment no, there isn't the option for that. There are a couple of issues:

1. I never thought of that before :)
2. The component needs to know about the Vue components so it knows when to invalidate them. That might be the biggest stumbling block here. It might require additional information to be passed in.

What's the use case for this over a named slot?

**Comment by AllanJard** [comment]
Hi - sorry for the delay in me getting back to you about this!

The appContext seems like the corner stone in this use case. My recent React integration for DataTables allows JSX to be returned in a column definition and I quite liked using that approach. I will have a look at this for Vue as well, although I'm not yet sure of the time scales as I've got a bunch of other stuff planned for development as well.

**Comment by ajmcgrail** [comment]
Hey there Allan!

I really appreciate the response! Take your time. And I agree that the appContext is very important, I think if that could be set correctly most other things would fall into place, but unfortunately I really wasn't able to find good documentation on *how* to do that. I don't think what I/you intend to do is a particularly common use case, unfortunately.

**Comment by db-hui** [comment]
Hi there! Do you plan to revisit this feature? Not having app context available in slots also means other libraries using a provide/inject pattern fail to load properly (e.g. i18n).

Thanks for working on this and publishing example code here!

**Comment by AllanJard** [comment]
Yes, very definitely plan to come back to this! I'm working on DataTables 3, which is basically taking all of my development time at the moment. I'm desperate to get it wrapped so I can turn to other things (such as improving this component). I'm knee deep in updating the docs at the moment, and then hope to have a beta.

So in short, yes, I absolutely want to come back to this and plan to do so, but I'm not yet certain when. Pull requests welcome!

**Comment by db-hui** [comment]
Thanks for these insights, looking forward to both DT3 and this feature!

**Comment by stanleyxu2005** [comment]
@ajmcgrail

I found your approach is smart and very advanced!
You ask `display` render to do nothing. When a cell is mounted. you create a vnode and mount to the cell.

I previously use manual slots to create a complex DT instance. But I found it will create 2 nested div elements around my cell component. I managed to reduce to 1 with [this PR](https://github.com/DataTables/Vue/pull/42)) but it is still not perfect, until I saw your approach.

Below is my simplified version using `createdCell`:
```javascript
export function myDataTableCell(columnProps = {}, CellComponent, toCellModel) {
  let defaultContent // can be set to other values, if data can be null, defaultContent must be defined!

  return {
    ...columnProps,

    data: toCellModel,
    defaultContent,

    render: {
      // Display render. if data is defined, set display content to empty string, otherwise default content.
      display: (el) => (el ? '' : defaultContent ), 

      // Other renders for sorting/ordering
      _: (cellData) => cellData ? cellData.getSomeValue() : '',
    },

    createdCell(cell, cellData, rowData, rowIndex, colIndex) {
      // If cellData is defined, create vnode and mount to TD
      const vnodeRequired = cellData && cellData !== defaultContent

      if (vnodeRequired) {
        const vnode = h(CellComponent, { cellData, /* ...other_params */ })

        render(vnode, cell)
        // Code not shown in this example:
        // (1) Register `cell` in a set (slightly different than the official impl of DtVue. I dont use tableId,row,col as key)
        // (2) Once table data is changed, find all `cell` that no longer in DOM and call `render(null, cell)` to trigger vnode unmounting and then remove registered `cell` from the set.
        //   see https://github.com/DataTables/Vue/pull/43
      }
    },
  }
}
```

I have some questions and suggestions to you and @AllanJard 

**Point 1:** Why you need a `watchEffect`? Do you really see these `vnode`s not being recycled? 
The original source code, all created host `<div>` will be deleted when it is not needed. There is no `render(null, elements[key])` nor `elements[key].remove()` and they cannot be, otherwise it will lead to errors.

**Point 2:** Why you need bind appContext to a component? Is it only very specific to your usage or it is useful in general. I would isolate Cell Component from any global context.

**Point 3:** I have extra handling for default content. If data is null, without default content, a warning dialog will popup. It's tricky to handle whether vnode is required, if the content is already default content.

Really happy to see this fanatic Vue implementation project and this valuable discussion here.

Best regards
S.X
